bofhound/ad/models/bloodhound_pki_template.py

```diff
@@ -0,0 +1,256 @@
+        'schemaversion', 'displayname', 'oid', 'enrollmentflag', 'requiresmanagerapproval',
+        'nosecurityextension', 'certificatenameflag', 'enrolleesuppliessubject', 'subjectaltrequireupn',
+        'subjectaltrequiredns', 'subjectaltrequiredomaindns', 'subjectaltrequireemail',
+        'subjectaltrequirespn', 'subjectrequireemail', 'ekus', 'certificateapplicationpolicy',
+        'authorizedsignatures', 'applicationpolicies', 'issuancepolicies', 'effectiveekus',
+        'authenticationenabled'
+    ]
+
+    def __init__(self, object):
+
+        super().__init__(object)
+
+        self._entry_type = "PKI Template"
+        self.GPLinks = []
+        self.ContainedBy = []
+        self.IsACLProtected = False
+        self.Properties["blocksinheritance"] = False
+        self.cas_ids = []
+
+        if 'objectguid' in object.keys():
+            self.ObjectIdentifier = object.get("objectguid")
+
+        if 'distinguishedname' in object.keys():
+            domain = ADUtils.ldap2domain(object.get('distinguishedname')).upper()
+            self.Properties['domain'] = domain
+            self.Properties['distinguishedname'] = object.get('distinguishedname').upper()
+
+        if 'description' in object.keys():
+            self.Properties['description'] = object.get('description')
+        else:
+            self.Properties['description'] = None
+
+        if 'pkiexpirationperiod' in object.keys():
+            pKIExpirationPeriod_b64 = object.get("pkiexpirationperiod")
+            pKIExpirationPeriod_byte_array = base64.b64decode(pKIExpirationPeriod_b64)
+            self.Properties["validityperiod"] = span_to_str(filetime_to_span(pKIExpirationPeriod_byte_array))
+
+        if 'pkioverlapperiod' in object.keys():
+            pKIRenewalPeriod_b64 = object.get("pkioverlapperiod")
+            pKIRenewalPeriod_byte_array = base64.b64decode(pKIRenewalPeriod_b64)
+            self.Properties["renewalperiod"] = span_to_str(filetime_to_span(pKIRenewalPeriod_byte_array))
+
+        if 'mspki-template-schema-version' in object.keys():
+            self.Properties['schemaversion'] = int(object.get('mspki-template-schema-version'))
+
+        if 'displayname' in object.keys():
+            self.Properties["displayname"] = object.get("displayname")
+            self.Properties["name"] = (object.get("displayname").replace(' ', '') + '@' + domain).upper()
+
+        if 'mspki-cert-template-oid' in object.keys():
+            self.Properties["oid"] = object.get("mspki-cert-template-oid")
+
+        if 'mspki-enrollment-flag' in object.keys():
+            enrollment_flag = object.get("mspki-enrollment-flag")
+            if enrollment_flag is not None:
+                enrollment_flag = MS_PKI_ENROLLMENT_FLAG(int(enrollment_flag))
+            else:
+                enrollment_flag = MS_PKI_ENROLLMENT_FLAG(0)
+            self.Properties["enrollmentflag"] = ', '.join(enrollment_flag.to_str_list())
```

Comment:
This class looks great, only difference I noticed between it's output and SharpHound's is that we're using Certipy's format of enrollment flag display (i.e., `AutoEnrollment, PublishToDs, PendAllRequests, IncludeSymmetricAlgorithms`) instead of the capitialized flag names SH uses (i.e., `INCLUDE_SYMMETRIC_ALGORITHMS, PEND_ALL_REQUESTS, PUBLISH_TO_DS, AUTO_ENROLLMENT`)

Not sure where those camel case are being pulled from, we're using the raw flag names for certificate name flags already and the code looks similar

Reply:
this comes from `cert_utils.py`:

```python
class IntFlag(enum.IntFlag):
    def to_list(self):
        cls = self.__class__
        members, _ = _decompose(cls, self._value_)
        return members

    def to_str_list(self):
        return list(map(lambda x: str(x), self.to_list()))

    def __str__(self):
        cls = self.__class__
        if self._name_ is not None:
            return "%s" % (to_pascal_case(self._name_))
        members, _ = _decompose(cls, self._value_)
        if len(members) == 1 and members[0]._name_ is None:
            return "%r" % (members[0]._value_)
        else:
            return "%s" % (
                ", ".join(
                    [to_pascal_case(str(m._name_ or m._value_)) for m in members]
                ),
            )

    def __repr__(self):
        return str(self)
```

Removing the `pascal_case()` should fix

Reply:
made a specific declaration of the `__str__` function for `MS_PKI_PRIVATE_KEY_FLAG` and `MS_PKI_ENROLLMENT_FLAG` classes